##### Shubham P Sharma

In [16]:
import pandas as pd
import numpy as np
import random
import math

In [17]:
#print(lines)

In [18]:
data = pd.read_table('3sat.txt', delim_whitespace=True, names=["A","B","C","o"])

In [19]:
data1=data[['A','B','C']]

In [20]:
#data1.head()


In [21]:
x=np.zeros(102)
#print(x)

            

In [22]:
for i in range (0,100):
    if(i%2 !=0):
        
        x[i] = 0
        
    else:
            
        x[i]=1
#for i in range(0,101):
    
    #print(i)
    #print(x[i])

In [23]:
data1['value'] =np.nan

In [24]:
def val(x,data1):
    
    y=0 
    for i in range(0,425):

            x1=data1.iloc[i,0]
            x2=data1.iloc[i,1]
            x3=data1.iloc[i,2]

            if(x1<0):
                x1=-1*x1

                v1=not x[x1]


            else:
                v1=x[x1]



            if(x2<0):
                x2=-1*x2
                v2= not x[x2]


            else:
                v2=x[x2]


            if(x3<0):
                x3=-1*x3
                v3=not x[x3]


            else:
                v3=x[x3]


            result=v1 or v2 or v3
            data1.loc[i,'value']=result
            
               
            x4=data1.iloc[i,3]
            if x4 == True :
                y = y+1

    #print(y)
    return y

In [25]:
def val111(x,data1):
    
    y=0 
    for i in range(0,425):

            x1=data1.iloc[i,0]
            x2=data1.iloc[i,1]
            x3=data1.iloc[i,2]

            if(x1<0):
                x1=-1*x1

                if x[x1-1] == '1':
                    v1=0
                else:
                    v1=1
                


            else:
                
                v1=int(x[x1-1])



            if(x2<0):
                x2=-1*x2
                if x[x2-1] == '1':
                    v2=0
                else:
                    v2=1
                


            else:
                v2=int(x[x2-1])

            if(x3<0):
                x3=-1*x3
                if x[x3-1] == '1':
                    v3=0
                else:
                    v3=1
                


            else:
                v3=int(x[x3-1])

            result=v1 or v2 or v3
            data1.loc[i,'value']=result
            
               
            x4=data1.iloc[i,3]
            if x4 == True :
                y = y+1

    #print(y)
    return y

#### HILL CLIMBING

In [12]:
z='1010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010'
#z='1111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111101000'

#print(len(z))
#print(int(z,2))
total1 = 0
total2 = 0
i=0
while(total1 <= total2):
    total1 = val111(z,data1)
    
    #print('total1',total1)
    added=0
    while(added ==0):
        added=random.randint(-1,1)
    
    z1 = str(bin(int(z,2) + added))[2:]
    total2 = val111(z1,data1)
    #print('total2',total2)
    if (total1<=total2):
        z=z1
    i+=1
#print(i)
    
#z
total = val111(z,data1)
print('achieved maximum',total)

#print(len(z))
print(int(z,2))

achieved maximum 373
845100400152152934331135470250


#### achieved maximum 373
#### input (base10) 845100400152152934331135470250

#### Simulated Annealing

In [13]:
def probability(change,temp):
    
    prob= math.exp((change)/temp)
    #print(prob)
    return prob



#### Neighbourhood Size (-1000, 1000) with 100 cooling 

In [14]:
initial_state='1010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010'
print('initial_state',int(initial_state,2))
total1=0
total2=0
temp=1

for i in range(0,100):
    total1=val111(initial_state,data1)
    #print('total1',total1)
    added=0
    while(added ==0):
        added=random.randint(-1000,1000)
        
    next_state=str(bin(int(z,2) + added))[2:]
    #print('next_state',int(next_state,2))


    total2=val111(next_state,data1)
    #print('total2',total2)
    
    change=total2-total1
   
        
    prob=probability(change,temp)
    
   
    def probadjustes(prob):

        if(prob>1):
            prob=1
        
    
        return prob
    
    probadj=probadjustes(prob)
    
    #print(probadj)
    
    variable_x=random.random()
    
    if(probadj>=variable_x):
            initial_state=next_state
            currentstate=initial_state
            
    temp=temp-0.01
    #print(temp)
    #print(i)

total_XC=val111(currentstate,data1)
#print(currentstate)
print('final state base 10',int(currentstate,2))
print('achieved maximum' ,total_XC)


initial_state 845100400152152934331135470250
final state base 10 845100400152152934331135470992
achieved maximum 379




#### initial_state 845100400152152934331135470250
#### final state base 10 845100400152152934331135470992
#### achieved maximum 379

#### Neighbourhood Size (-500, 500) with 1500 cooling 

In [26]:
initial_state='1010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010101010'
print('initial_state',int(initial_state,2))
total1=0
total2=0
temp=1

for i in range(0,1500):

    total1=val111(initial_state,data1)
    #print('total1',total1)
    added=0
    while(added ==0):
        added=random.randint(-500,500)
        
    next_state=str(bin(int(z,2) + added))[2:]
    #print('next_state',int(next_state,2))


    total2=val111(next_state,data1)
    #print('total2',total2)
    
    change=total2-total1
   
        
    prob=probability(change,temp)
    
   
    def probadjustes(prob):

        if(prob>1):
            prob=1
        
    
        return prob
    
    probadj=probadjustes(prob)
    
    #print(probadj)
    
    variable_x=random.random()
    
    if(probadj>=variable_x):
            initial_state=next_state
            currentstate=initial_state
            
    temp=temp-(1/1500)
    #print(temp)
    #print(i)

total_XC=val111(currentstate,data1)
#print(currentstate)
print('final state base 10',int(currentstate,2))
print('achieved maximum',total_XC)



initial_state 845100400152152934331135470250
final state base 10 845100400152152934331135469840
achieved maximum 380


 #### initial_state 845100400152152934331135470250
#### final state base 10 845100400152152934331135469840
##### achieved maximum 380 